kanidmd/src/lib/valueset.rs

```diff
@@ -1,112 +1,1297 @@
+use crate::credential::Credential;
 use crate::prelude::*;
-// use hashbrown::HashSet;
-use std::borrow::Borrow;
-use std::collections::BTreeSet;
+use crate::repl::cid::Cid;
+use either::Either::{Left, Right};
+use kanidm_proto::v1::Filter as ProtoFilter;
+use smolset::{SmolSet, SmolSetIter};
+use sshkeys::PublicKey as SshPublicKey;
+use std::collections::btree_map::Entry as BTreeEntry;
+use std::collections::{BTreeMap, BTreeSet};
 use std::iter::FromIterator;
+use time::OffsetDateTime;
+use tracing::trace;
+
+use crate::be::dbvalue::{
+    DbCidV1, DbValueCredV1, DbValueEmailAddressV1, DbValueTaggedStringV1, DbValueV1,
+};
+use crate::value::DataValue;
+
+// SmolSet<[; 1]>
+
+#[derive(Debug, Clone)]
+enum I {
+    Utf8(BTreeSet<String>),
+    Iutf8(BTreeSet<String>),
+    // Could be AttrString?
+    Iname(BTreeSet<String>),
+    Uuid(BTreeSet<Uuid>),
+    Bool(SmolSet<[bool; 1]>),
+    Syntax(SmolSet<[SyntaxType; 1]>),
+    Index(SmolSet<[IndexType; 1]>),
+    Refer(BTreeSet<Uuid>),
+    JsonFilt(SmolSet<[ProtoFilter; 1]>),
+    Cred(BTreeMap<String, Credential>),
+    SshKey(BTreeMap<String, String>),
+    SecretValue(SmolSet<[String; 1]>),
+    Spn(BTreeSet<(String, String)>),
+    Uint32(SmolSet<[u32; 1]>),
+    Cid(SmolSet<[Cid; 1]>),
+    Nsuniqueid(BTreeSet<String>),
+    DateTime(SmolSet<[OffsetDateTime; 1]>),
+    EmailAddress(BTreeSet<String>),
+    Url(SmolSet<[Url; 1]>),
+}
 
 pub struct ValueSet {
-    inner: BTreeSet<Value>,
+    inner: I,
+}
+
+macro_rules! mergesets {
+    (
+        $a:expr,
+        $b:expr
+    ) => {{
+        $b.iter().for_each(|v| {
+            if !$a.contains(v) {
+                $a.insert(v.clone());
+            }
+        });
+        Ok(())
+    }};
 }
```

Comment:
Because we have to be able to merge across btreeset and smolset, and the type signature for that as a function would be horrific.

I think you are right about the contains / insert though, so I'll change that.